src/core/sock.c

```diff
@@ -1088,3 +1161,24 @@ void nn_sock_stat_increment (struct nn_sock *self, int name, int64_t increment)
             break;
     }
 }
+
+int nn_sock_hold (struct nn_sock *self)
+{
+    switch (self->state) {
+    case NN_SOCK_STATE_ACTIVE:
+    case NN_SOCK_STATE_INIT:
+        self->holds++;
+        return 0;
+    case NN_SOCK_STATE_ZOMBIE:
+        return -ETERM;
+    }
```

Comment:
I like this strategy.

And on the record, I feel that the @sustrik design decision to put `assert()` all over the place was absolutely correct and has made for a better library. That's another topic ... but to tie into what you say, libnanomsg is definitely at the tipping point nearing v1.0 where `assert()` should no longer be sprinkled, but a run-time error code instead.

That said, I'd be in favor of a new error code -- say ENANOMSG or such -- whose existence in the field would be distinct from EBADF.

The semantic of ENANOMSG alerts nanomsg library devs that something is wrong and needs to be fixed, whereas EBADF alerts library users (application developers) that they have made a logic error.

This could tie in with work @TTimo pushed for long ago, turning `nn_assert` into a logger rather than a panic stop.

Thoughts?